internal/driver/tagroot.go

```diff
@@ -0,0 +1,117 @@
+package driver
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/google/pprof/profile"
+)
+
+// addLabelNodes adds psuedo-stack-frames "label:value" to each Sample with
+// labels matching the supplied keys.
+//
+// rootKeys adds frames at the root of the callgraph (first key becomes new root).
+// leafKeys adds frames at the leaf of the callgraph (last key becomes new leaf).
+//
+// Returns whether there were matches found for the label keys.
+func addLabelNodes(p *profile.Profile, rootKeys, leafKeys []string) (rootm, leafm bool) {
+	// Find where to insert the new locations and functions at the end of
+	// their ID spaces.
+	var maxLocID uint64
+	var maxFunctionID uint64
+	for _, loc := range p.Location {
+		if loc.ID > maxLocID {
+			maxLocID = loc.ID
+		}
+	}
+	for _, f := range p.Function {
+		if f.ID > maxFunctionID {
+			maxFunctionID = f.ID
+		}
+	}
+	nextLocID := maxLocID + 1
+	nextFuncID := maxFunctionID + 1
+
+	// Intern the new frames we are generating.
+	locsByName := map[string]*profile.Location{}
+	functionsByName := map[string]*profile.Function{}
+
+	internFunction := func(functionName string) *profile.Function {
+		function, found := functionsByName[functionName]
+		if found {
+			return function
+		}
+		function = &profile.Function{
+			ID:   nextFuncID,
+			Name: functionName,
+		}
+		nextFuncID++
+		p.Function = append(p.Function, function)
+		functionsByName[functionName] = function
+		return function
+	}
+
+	internLocation := func(functionName string, function *profile.Function) *profile.Location {
+		location, found := locsByName[functionName]
+		if found {
+			return location
+		}
+		location = &profile.Location{
+			ID: nextLocID,
+			Line: []profile.Line{
+				profile.Line{
+					Function: function,
+				},
+			},
+		}
+		nextLocID++
+		p.Location = append(p.Location, location)
+		locsByName[functionName] = location
+		return location
+	}
+
+	for _, s := range p.Sample {
+		rootsToAdd := []*profile.Location{}
+		leafsToAdd := []*profile.Location{}
+		for i := range rootKeys {
+			// Loop backwards, ensuring the first tag is the new root.
+			k := rootKeys[len(rootKeys)-1-i]
+			values := s.Label[k]
+			if len(values) > 0 {
+				rootm = true
+			}
+			functionName := k + ":" + strings.Join(values, ",")
+			function := internFunction(functionName)
+			location := internLocation(functionName, function)
+			rootsToAdd = append(rootsToAdd, location)
+		}
+		for i, _ := range leafKeys {
+			// Loop backwards, ensuring the last tag is the new leaf.
+			k := leafKeys[len(leafKeys)-1-i]
+			values := []string{}
+			values = append(values, s.Label[k]...)
+			numLabels := s.NumLabel[k]
+			numUnits := s.NumUnit[k]
+			if len(numLabels) == len(numUnits) {
+				for i, numLabel := range numLabels {
+					unit := numUnits[i]
+					values = append(values, fmt.Sprintf("%d%s", numLabel, unit))
```

Comment:
Unfortunately, there is one more piece of complexity you'll have to deal with: I think we should scale the numeric tag values in the same way we scale them when printing in other circumstances, such as `printTags` in report.go: see [code](https://github.com/google/pprof/blob/02619b876842e0d0afb5e5580d3a374dad740edb/internal/report/report.go#L688-L690).

Reply:
Thanks for the pointer, very helpful. Done